File: src/main/java/com/csoft/gluon/spi/Router.java
package com.csoft.gluon.spi;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class Router {

    private static final Logger LOGGER = LoggerFactory.getLogger(Router.class);

    private Map<String, Controller> gets = new HashMap<>();
    private Map<String, Controller> deletes = new HashMap<>();
    private Map<String, Controller> posts = new HashMap<>();
    private Map<String, Controller> puts = new HashMap<>();

    private Router() {
    }

    public static Router newGluonRouter() {
        return new Router();
    }

    private static void addMethodToSet(Map<String, Set<String>> map, String key, String method) {
        Set<String> value = map.getOrDefault(key, new HashSet<>());
        value.add(method);
        map.put(key, value);
    }

    private static void addControllerToSet(Map<String, Set<Controller>> map, String key, Controller controller) {
        Set<Controller> value = map.getOrDefault(key, new HashSet<>());
        value.add(controller);
        map.put(key, value);
    }

    public Router get(final String path, final Controller controller) {
        LOGGER.info("Registering path [{}] with [method=GET, controller={}]", path, controller.getClass().getCanonicalName());
        gets.put(path, controller);
        return this;
    }

    public Router post(final String path, final Controller controller) {
        LOGGER.info("Registering path [{}] with [method=POST, controller={}]", path, controller.getClass().getCanonicalName());
        posts.put(path, controller);
        return this;
    }

    public Router put(final String path, final Controller controller) {
        LOGGER.info("Registering path [{}] with [method=PUT, controller={}]", path, controller.getClass().getCanonicalName());
        puts.put(path, controller);
        return this;
    }

    public Router delete(final String path, final Controller controller) {
        LOGGER.info("Registering path [{}] with [method=DELETE, controller={}]", path, controller.getClass().getCanonicalName());
        deletes.put(path, controller);
        return this;
    }

    public Map<String, Set<String>> mapOfMethods() {
        Map<String, Set<String>> map = new HashMap<>();
        gets.forEach((k, v) -> addMethodToSet(map, k, "GET"));
        posts.forEach((k, v) -> addMethodToSet(map, k, "POST"));
        puts.forEach((k, v) -> addMethodToSet(map, k, "PUT"));
        deletes.forEach((k, v) -> addMethodToSet(map, k, "DELETE"));
        return map;
    }

    public Map<String, Set<Controller>> mapOfControllers() {
        Map<String, Set<Controller>> map = new HashMap<>();
        gets.forEach((k, v) -> addControllerToSet(map, k, v));
        posts.forEach((k, v) -> addControllerToSet(map, k, v));
        puts.forEach((k, v) -> addControllerToSet(map, k, v));
        deletes.forEach((k, v) -> addControllerToSet(map, k, v));
        return map;
    }

}

File: src/main/kotlin/com/csoft/gluon/spi/Router.kt
package com.csoft.gluon.spi

import mu.KotlinLogging
import java.util.*

class Router {

    private val logger = KotlinLogging.logger {}

    private val gets = HashMap<String, Controller>()
    private val deletes = HashMap<String, Controller>()
    private val posts = HashMap<String, Controller>()
    private val puts = HashMap<String, Controller>()

    operator fun get(path: String, controller: Controller): Router {
        logger.info("Registering path [{}] with [method=GET, controller={}]", path, controller.javaClass.canonicalName)
        gets[path] = controller
        return this
    }

    fun post(path: String, controller: Controller): Router {
        logger.info("Registering path [{}] with [method=POST, controller={}]", path, controller.javaClass.canonicalName)
        posts[path] = controller
        return this
    }

    fun put(path: String, controller: Controller): Router {
        logger.info("Registering path [{}] with [method=PUT, controller={}]", path, controller.javaClass.canonicalName)
        puts[path] = controller
        return this
    }

    fun delete(path: String, controller: Controller): Router {
        logger.info("Registering path [{}] with [method=DELETE, controller={}]", path, controller.javaClass.canonicalName)
        deletes[path] = controller
        return this
    }

    fun mapOfMethods(): Map<String, Set<String>> {
        val map = HashMap<String, Set<String>>()
        gets.forEach { k, _ -> addMethodToSet(map, k, "GET") }
        posts.forEach { k, _ -> addMethodToSet(map, k, "POST") }
        puts.forEach { k, _ -> addMethodToSet(map, k, "PUT") }
        deletes.forEach { k, _ -> addMethodToSet(map, k, "DELETE") }
        return map
    }

    fun mapOfControllers(): Map<String, Set<Controller>> {
        val map = HashMap<String, Set<Controller>>()
        gets.forEach { k, v -> addControllerToSet(map, k, v) }
        posts.forEach { k, v -> addControllerToSet(map, k, v) }
        puts.forEach { k, v -> addControllerToSet(map, k, v) }
        deletes.forEach { k, v -> addControllerToSet(map, k, v) }
        return map
    }

}

private fun addMethodToSet(map: MutableMap<String, Set<String>>, key: String, method: String) {
    val value = map.getOrDefault(key, HashSet()).plus(method)
    map[key] = value
}

private fun addControllerToSet(map: MutableMap<String, Set<Controller>>, key: String, controller: Controller) {
    val value = map.getOrDefault(key, HashSet()).plus(controller)
    map[key] = value
}
